buzzmobile/process/bearing/bearing.py

```diff
@@ -25,31 +25,71 @@ def median(self):
             return (sortedArray[index] + sortedArray[index + 1]) / 2.0
 
 
-#Global Variables
-bearings = {}
-bearings['last_fix'] = None
-bearings['med_filter'] = MedianFilter(rospy.get_param('median_filter_size'))
+# Global variables
+globalVars = {}
+globalVars['last_fix'] = None
+globalVars['med_filter'] = MedianFilter(rospy.get_param('median_filter_size'))
 bearing_pub = rospy.Publisher('bearing', Float64, queue_size=1)
 
+MIN_FIX_DISTANCE = rospy.get_param('min_fix_distance')
+EARTH_RADIUS = 6.3710088e6
+
+
 def bearing(fix):
-    global last_fix
-    global med_filter
-
-    if bearings['last_fix'] is not None:
-        lat1 = math.radians(bearings['last_fix'].latitude)
-        lon1 = math.radians(bearings['last_fix'].longitude)
-        lat2 = math.radians(fix.latitude)
-        lon2 = math.radians(fix.longitude)
-        y = math.sin(lon2 - lon1) * math.cos(lat2)
-        x = math.cos(lat1)*math.sin(lat2) - math.sin(lat1)*math.cos(lat2)*math.cos(lon2 - lon1)
-        angle = math.atan2(y, x)
-        computedBearing = (angle + 2*math.pi) % 2*math.pi
-        bearings['med_filter'].add(computedBearing)
-
-        bearing_pub.publish(bearings['med_filter'].median())
+    """
+    Adds the latest computed bearing to the meadian filter, publishes
+    the current median bearing, and sets last_fix to the given fix
+    if the given fix is >= MIN_FIX_DISTANCE from last_fix or there is
+    currently no last_fix
+    """
+
+    last_fix = globalVars['last_fix']
+    med_filter = globalVars['med_filter']
 
     if fix is not None:
-        last_fix = fix
+        if last_fix is not None:
+            distance = get_distance(last_fix, fix)
+            bearing = get_forward_angle(last_fix, fix)
+            med_filter.add(bearing)
+            bearing_pub.publish(med_filter.median())
+            if distance >= MIN_FIX_DISTANCE:
+                last_fix = fix
+        else:
+            last_fix = fix
+    globalVars['last_fix'] = last_fix
```

Comment:
rapha no!

Its a scoping rule. If a value is ever assigned inside a function, it is assigned a local varname and a value (see `some_method.__code__.co_varnames`). If it is never assigned to, then a recursive lookup is used to parent (nonlocal) function and finally the global scope.

That's why the result of 

    a = 5
    def func():
        a = 7
    func()
    print(a)

will be 5, and not 7, and also why

    a = 5
    def f():
        b = a
        a = 5
    f()

will raise UnboundLocalError. Depsite the fact that `a` exists at runtime, at compile time (that is, when the function was compiled), it was created as a local. And that local isn't yet bound in the function.

To fix, use the `global` keyword.

    a_global
    def message_cb(new_msg):
        global a_global
        a_global = new_msg

This instructs python to use the global version of `a_global` in this function instead of creating a local.